### ISSUE TITLE
Guidelines for informative error messages

### ISSUE THREAD

**Issue by ngriffiths21** [tidyverse/style#157]
In section 9.1 on error messages, the guide says:

> Use `stop(call. = FALSE), rlang::abort(), Rf_errorcall(R_NilValue, ...)` to avoid cluttering the error message with the name of the function that generated it. That information is often not informative, and can easily be accessed via traceback() or an IDE equivalent.

I disagree that the function name is uninformative, and just wanted to share my two cents on error handling.

I bet many would agree that error handling is one of the most painful parts of working with R. Debugging can take lots of time, and it seems like a really valuable thing to try to improve. In my own experience using packages, the biggest challenge has always been understanding the *context* of the error rather than the nature of the error itself. I'll show what I mean with a simple example, one that follows the guidelines but still results in a confusing error.

In the simplest case, we call a function interactively and get a directly relevant message. It's easy to see what's going on here:

``` r
library(tidyverse)
select(mtcars, nonexistent)
#> Error: Can't subset columns that don't exist.
#> ✖ Column `nonexistent` doesn't exist.
```

<sup>Created on 2020-07-26 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

However, the far more common situation is that the error was thrown from deeper in the call stack:

``` r
obscure_fn()
#> Error: Can't subset columns that don't exist.
#> ✖ Column `nonexistent` doesn't exist.
```

<sup>Created on 2020-07-26 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

This error message is completely uninformative. To understand it, you need to find out where the error is being thrown, so the function name is arguably more important than the message. Unfortunately (and despite what the guide says), the offending function call is *not* easily identifiable from the trace.

``` r
rlang::last_error()
#> <error/vctrs_error_subscript_oob>
#> Can't subset columns that don't exist.
#> ✖ Column `nonexistent` doesn't exist.
#> Backtrace:
#>   1. global::obscure_fn()
#>  22. vctrs:::stop_subscript_oob(...)
#>  23. vctrs:::stop_subscript(...)
#> Run `rlang::last_trace()` to see the full context.
```

The trace claims that `vctrs:::stop_subscript` threw the error, which isn't wrong, but it's misleading. I think the error gets handled in `tidyselect:::eval_select_impl`, but I can only tell by reading the source. If that's true, the error message could easily include "Error in tidyselect::eval_select" along with the other notes. Note that this message plus the `tidyselect::eval_select` documentation are sufficient information to understand exactly what went wrong: I passed some dataframe to `eval_select()`, and it didn't contain the `nonexistent` column.

To actually fix the problem, I need to identify which line in *my* code resulted in selecting the wrong column. This currently requires looking around in the trace. In this (overly simplistic) example, I just need to identify the line `dplyr::select(mtcars, nonexistent)`, which is called from `obscure_fn()`. Since the trace can become cluttered with internal package calls, magrittr/purrr calls, etc, I bet there are ways to improve this aspect of debugging as well.

To summarize:

- I believe it's important to identify the function in the error message. To really understand an error, you need to know which function threw it and what that function does.
- The traceback can get quite complicated with error handling and other calls, and I disagree that contextual info can generally be easily accessed with `traceback()`
- Before you can fix the bug, you need to find the code you wrote that led to the error. This is probably the more complicated problem and maybe out of scope. Is there currently a way to distinguish between the trace from local code &mdash; which is both more familiar and more likely to contain the bug &mdash; versus package code? Even better, could an error message report the earliest package function call on the path to the error?

**Comment by luciorq** [comment]
Surely it would be beneficial to know the depth of the origin of the error. Being able to differentiate internal and surface errors would be interesting.

**Comment by ngriffiths21** [comment]
> But it need to be a conditional thing, since every function could be used on the "surface".

That's interesting, there could be some creative ways to make error handling a little nicer.

"Internal" errors seem common (whether because a user wrote their own function, or because package code didn't handle the error appropriately), and that type is also harder to debug. So I don't think error messages should ever assume the error is on the surface, and should still make some sense if it isn't.

**Comment by hadley** [comment]
We're moving towards displaying the function name (but not full call) in the error: https://github.com/r-lib/rlang/pull/1242.

I'll make a note that we need to update our error message advice after that's released.
